Fix invalid pathname slashes in "getURL()" utility

takueof opened the pull request:
## For Contributors

### Improving Documentation

- [x] Run `pnpm prettier-fix` to fix formatting issues before opening the PR.
- [x] Read the Docs Contribution Guide to ensure your contribution follows the docs guidelines: https://nextjs.org/docs/community/contribution-guide

### Adding or Updating Examples

- [x] The "examples guidelines" are followed from our contributing doc https://github.com/vercel/next.js/blob/canary/contributing/examples/adding-examples.md
- [x] Make sure the linting passes by running `pnpm build && pnpm lint`. See https://github.com/vercel/next.js/blob/canary/contributing/repository/linting.md

### Fixing a bug

- [ ] Related issues linked using `fixes #number`
- [x] Tests added. See: https://github.com/vercel/next.js/blob/canary/contributing/core/testing.md#writing-tests-for-nextjs
- [x] Errors have a helpful link attached, see https://github.com/vercel/next.js/blob/canary/contributing.md

## For Maintainers

- [ ] Minimal description (aim for explaining to someone not on the team to understand the PR)
- [ ] When linking to a Slack thread, you might want to share details of the conclusion
- [ ] Link both the Linear (Fixes NEXT-xxx) and the GitHub issues
- [ ] Add review comments if necessary to explain to the reviewer the logic behind a change

### What?

If you access a root page that has two or more extra slashes at the end, such as [https://foo.com//](https://foo.com//), errors throw in the `URL` constructor and Next.js will stop rendering.

### Why?

Because [`getURL()`](https://github.com/vercel/next.js/blob/d36917cd3465a03b6f6a85d52f58e2be172d55a9/packages/next/src/shared/lib/utils.ts#L332-L336) utility function in `/packages/next/src/shared/lib/utils.ts` does not remove two or more extra slashes.

### How?

Add processing to the `getURL()` utility function to replace two or more leading slashes with one slash.

## Note

- This bug [has been around for over 7 years](https://github.com/vercel/next.js/blame/96fca5e2ad1f2fcf368dcbc2ad2cab8ae949b0e0/lib/utils.js#L62-L66) since the PR was created
- I created a PR because this bug can't be fixed without addressing it on the Next.js side

Closes NEXT-
Fixes #

ijjk commented:
Allow CI Workflow Run

- [ ] approve CI run for commit: 393bb14b128fdccb4810dc4f2a77fe864b0c9a59

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

takueof commented:
What's happening with CI?

Even though the local self-check was fine 🤔 

![screenshot_ci_error_01](https://github.com/user-attachments/assets/6d1e7f33-afe9-428a-a00f-b278861b9d23)

![screenshot_ci_error_02](https://github.com/user-attachments/assets/81104226-c84f-41c5-9710-99bfe811358e)

![screenshot_ci_error_03](https://github.com/user-attachments/assets/5b3dbef0-ac4c-4210-b6cf-7b8994c53196)

huozhi commented:
If you have more trailing slashes, even standard util will treat them as part of pathname, sounds correct it was thrown as invalid URL?

```
new URL('https://fooo.com//').pathname
>> '//'
```

takueof commented:
@huozhi 
Thank you for your reply.

Yes, but it doesn't happen in your example...

It occurs with the following pattern (this is the code pattern used within [`/packages/next/src/shared/lib/router/utils/as-path-to-search-params.ts`](https://github.com/vercel/next.js/blob/528700d7b5813ce9467a6e41a9cb4e4165c3c274/packages/next/src/shared/lib/router/utils/as-path-to-search-params.ts#L4)):

```JavaScript
new URL(asPath, 'http://n').searchParams
```

The following error should be thrown:

```
Uncaught TypeError: Failed to construct 'URL': Invalid URL
```

Screenshot:

<img width="800" alt="new URL() throw by Google Chrome in JavaScript console" src="https://github.com/user-attachments/assets/266aa801-115c-4adc-ba6f-c710db656fa0">

`V8` (Blink), `SpiderMonkey` (Gecko) and `JavaScriptCore` (WebKit) all have the same behavior.

takueof commented:
Initially, I thought that changing

```JavaScript
new URL(asPath, 'https://n.com').searchParams
```

to

```JavaScript
new URL(`https://n.com${asPath}`).searchParams
```

would complete the fix.

However, when I investigated deeper, I started to think that the fundamental solution might not be solved unless I changed `getURL()`.

The reason is that the return value of `getURL()` is different for `http://` and `https://`.

- If we execute now `getURL()` on `http://n.com//`, `/` will be returned
- If we execute now `getURL()` on `https://n.com//`, `//` will be returned

By fixing the core part, we can avoid problems that may occur in other code that uses the result of `getURL()`.

huozhi commented:
The test seem cannot demostrate the issue you describe, if I manually run in https mode and just visit the page, it looks working correctly 🤔

takueof commented:
@huozhi 
Thank you for checking.

The e2e tests I wrote were working fine as of a559e15.
However, after repeated `rebase`, the e2e test itself fails as of 23fc909...

```shell
pnpm build && pnpm test-start test/e2e/invalid-url-slash-https/invalid-url-slash-https.test.ts
```

The error that occurs at that time is a part that I did not write, and I don't understand the cause of the error:

```shell
      24 |   }
      25 |
    > 26 |   await execa('pnpm', args, {
         |   ^
      27 |     cwd,
      28 |     stdio: ['ignore', 'inherit', 'inherit'],
      29 |     env: process.env,

      at makeError (../node_modules/.pnpm/execa@2.0.3/node_modules/execa/lib/error.js:58:11)
      at handlePromise (../node_modules/.pnpm/execa@2.0.3/node_modules/execa/index.js:112:26)
      at installDependencies (lib/create-next-install.js:26:3)
      at Span.traceAsyncFn (../packages/next/src/trace/trace.ts:147:14)
      at lib/create-next-install.js:166:9
      at Span.traceAsyncFn (../packages/next/src/trace/trace.ts:147:14)
      at createNextInstall (lib/create-next-install.js:44:10)
      at lib/next-modes/base.ts:223:36
      at Span.traceAsyncFn (../packages/next/src/trace/trace.ts:147:14)
      at NextStartInstance.createTestDir (lib/next-modes/base.ts:148:5)
      at NextStartInstance.setup (lib/next-modes/next-start.ts:23:5)
      at lib/e2e-utils/index.ts:201:7
      at Span.traceAsyncFn (../packages/next/src/trace/trace.ts:147:14)
      at createNext (lib/e2e-utils/index.ts:166:12)
      at Object.<anonymous> (lib/e2e-utils/index.ts:255:14)
```

Hmm, Have the e2e test specifications changed...? 🤔

I'll rewrite it if it changes, so I'm sorry to trouble you, but do you know the specifications that have changed?

takueof commented:
Oh, after splitting the test file into `http` and `https`, the e2e test is now working.

If you `git pull` the latest commit and run the following command, it does what I want:

```shell
pnpm build && pnpm test-start test/e2e/invalid-url-slash
```

I'm very sorry.

huozhi commented:
Tests are passed regardless we have the changes

takueof commented:
@huozhi 
Thanks to reply.

I also confirmed that the e2e test did not fail using the old code.

Then I remembered that the case I encountered was a `Pages Router`, not an `App Router`.

When I rewrote the e2e test using `Pages Router`, changed `utils.ts` to its pre-patch state, and checked again, the e2e test failed.

After applying the patch to `utils.ts`, the e2e test now passes.

Thank you for making me realize that.

We have pushed the latest code, so please test it yourself 🙏

takueof commented:
Hello, maintainers!
Thank you for considering my PR.

It's been about 4 weeks since I last commented.
If you don't mind, would it be possible for you to tell me about your current review status...?

Sorry for the inconvenience 🙏

ijjk commented:
## Failing test suites <!--29567481652-->
Commit: 6a78b62a79d179143fccd81ab3be1919fc959777

`pnpm test-dev test/development/acceptance-app/undefined-default-export.test.ts`
- Undefined default export > should error when page component export is not valid

<details>
<summary>Expand output</summary>

  ● Undefined default export › should error when page component export is not valid

    Expected Redbox but found none

      61 |     )
      62 |
    > 63 |     await session.assertHasRedbox()
         |     ^
      64 |     expect(await session.getRedboxDescription()).toInclude(
      65 |       'The default export is not a React Component in "/server-with-errors/page-export/page"'
      66 |     )

      at Object.<anonymous> (development/acceptance-app/undefined-default-export.test.ts:63:5)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

takueof commented:
@ijjk 
Thank you for your comment.

I didn't expect this PR at this point to have such an impact on other features (testing).
I will create a patch PR in my local environment while doing full testing again.

I apologize for taking up your valuable time.

takueof commented:
@ijjk (Cc: @huozhi)
Hello.
Thank you for letting me know about the test failure the other day.

After that, I reviewed the patch and changed the patch so that the existing test set would not fail and the new test set would pass.

I apologize for the inconvenience, but I would appreciate it if you could review the PR again 🙏

The following screenshots are the results of the retest:

## `pnpm test-dev test/development/acceptance-app/undefined-default-export.test.ts`

<img width="600" alt="" src="https://github.com/user-attachments/assets/2bc9d78c-50b9-41e8-b51e-f6e6b8339af2">

## `pnpm test test/integration/repeated-slashes/test/index.test.js`

<img width="600" alt="" src="https://github.com/user-attachments/assets/078da649-cdce-4ef6-88be-572a9ef43d23">

takueof commented:
Hello.
I tried running the same tests as CI in my local environment, but the following command seems to give unintended results.

```shell
pnpm test-start test
```

I looked at [the testing section](https://github.com/vercel/next.js/blob/882d88381c12a794e82afccc118435889dedd96f/contributing/core/testing.md) in the documentation, but there was **no command written to run exhaustive tests equivalent to CI**.

I apologize for the inconvenience, but would it be possible for you to give me a list of commands (I believe there are probably multiple) to run the same tests in the local environment as those running in CI...?

takueof commented:
Hello, can anyone answer the above question...?

takueof commented:
@ijjk @huozhi 

Please Help me.

Is it possible to read [a comment mentioned above](https://github.com/vercel/next.js/pull/70144?new_mergebox=true#issuecomment-2466514169) and receive a reply...?
